url90b6689baa_adsonrocha_PD321_tgz-pJ8-MyFuzzyLiteClassJ8/src/com/fuzzylite/rule/Antecedent.java

```diff
@@ -0,0 +1,271 @@
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.StringTokenizer;
+
+/**
+ *
+ * @author jcrada
+ */
+public class Antecedent {
+
+    protected Expression root;
+
+    public Expression getRoot() {
+        return this.root;
+    }
+
+    public double activationDegree(TNorm conjunction, SNorm disjunction) {
+        return this.activationDegree(conjunction, disjunction, root);
+    }
+
+    public double activationDegree(TNorm conjunction, SNorm disjunction, Expression node) {
+        if (node instanceof Proposition) {
+            Proposition proposition = (Proposition) node;
+            if (!proposition.variable.isEnabled()) {
+                return 0.0;
+            }
+            if (!proposition.getHedges().isEmpty()) {
+                int lastIndex = proposition.getHedges().size() - 1;
+                Hedge any = proposition.getHedges().get(lastIndex);
+                if (any instanceof Any) {
+                    double result = any.hedge(Double.NaN);
+                    while (--lastIndex >= 0) {
+                        result = proposition.getHedges().get(lastIndex).hedge(result);
+                    }
+                    return result;
+                }
+
+            }
+
+            InputVariable inputVariable = (InputVariable) proposition.getVariable();
+            double result = proposition.getTerm().membership(inputVariable.getInputValue());
+            for (int i =  proposition.getHedges().size() - 1 ;i >= 0; --i) {
+                result = proposition.getHedges().get(i).hedge(result);
+            }
+            return result;
+        } else if (node instanceof Operator) {
+            Operator operator = (Operator) node;
+            if (operator.getLeft() == null || operator.getRight() == null) {
+                throw new RuntimeException("[syntax error] left and right operators cannot be null");
+            }
+            if (Rule.FL_AND.equals(operator.getName())) {
+                return conjunction.compute(
+                        activationDegree(conjunction, disjunction, operator.getLeft()),
+                        activationDegree(conjunction, disjunction, operator.getRight()));
+            }
+            if (Rule.FL_OR.equals(operator.getName())) {
+                return disjunction.compute(
+                        activationDegree(conjunction, disjunction, operator.getLeft()),
+                        activationDegree(conjunction, disjunction, operator.getRight()));
+            }
+            throw new RuntimeException(String.format(
+                    "[syntax error] operator <%s> not recognized",
+                    operator.getName()));
+        } else {
+            throw new RuntimeException("[expression error] unknown instance of Expression");
+        }
+    }
+
+    public void load(String antecedent, Engine engine) {
+        Function function = new Function();
+        String postfix = function.toPostfix(antecedent);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;antecedent&#39; (index: 0 | type: String). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;engine&#39;. The same parameter type is validated here for example: url90b6689baa_adsonrocha_PD321_tgz-pJ8-MyFuzzyLiteClassJ8/src/com/fuzzylite/Console.java:263. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).